File: extras/python/psycho/levitt_experiment.py
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

"""Python code implementing the Levitt up-down transformed experiment protocol.

See Harry Levitt's paper for more details:
  http://bdml.stanford.edu/twiki/pub/Haptics/DetectionThreshold/psychoacoustics.pdf

Note, many of the examples here use the data from his Figure 4.
"""

import operator
from typing import Any, Dict
import matplotlib.pyplot as plt
import numpy as np


class LevittExp(object):
  """Implement the basic Levitt transformed up-down experiment."""

  def __init__(self,
               initial_level: float,
               change_delta: float = 1,
               decrease_step_by_run: bool = False,
               multiplicative_step: float = False,
               debug: bool = False):
    """Implements an up-down experiment searching for a threshold.

    This class takes in true/false decisions by the user and changes the
    level for the next experiment.  After creating the class with the
    appropriate parameters, call note_response() with each new user action.
    This method notes the response, and changes the level based on the
    internal response_rule() method.  The new level is returned for use in
    the next trial.

    Args:
      initial_level: Initial value for the experiment.
      change_delta: How much to change the level after each trial. When
        multiplicative_step is False, then the current step is
        add/subtracted from the current level. When multiplicative_step is
        True, then the current level is multiplied/divided by 1 plus the current
        step.
      decrease_step_by_run: The current step changes after each run (a
        monotonic change in the level).  As Levitt suggests, the current
        step is the initial_step divided by the number of runs so far.
      multiplicative_step: When true, the delta value represents a
          multiplicative factor, while when False the steps are additive.
        debug: Prints some internal state after each trial.
    """
    if change_delta < 0:
      raise ValueError('Change_delta should be positive, not %g.' %
                       change_delta)
    if multiplicative_step and change_delta > 1:
      raise ValueError('Multiplicative change should be less than 1, not %g.' %
                       change_delta)
    if multiplicative_step and initial_level == 0:
      raise ValueError('Do not start multiplicative steps starting at 0.')
    self.initial_level = initial_level
    self.level = initial_level
    self.initial_change_delta = change_delta
    self.change_delta = change_delta
    self.run_trial_history = [[]]  # A nested history of trials (run x trial #)
    self.change_history = [
    ]  # Keep track of change events so we can delineate runs
    self.level_history = []  # History of all levels over time
    self.trial_number = 0
    self.decrease_step_by_run = decrease_step_by_run
    self.multiplicative_step = multiplicative_step
    self.debug = debug

  @property
  def run_number(self) -> int:
    return len(self.run_trial_history)

  def note_response(self, new_answer):
    """Records the response to a trial, adjusts history and sets new level."""
    self.level_history.append(self.level)
    self.run_trial_history[-1].append(new_answer)
    self._response_rule()
    self.trial_number += 1
    if self.debug:
      last_run = self.run_trial_history[-1]
      new_trials = len(last_run)
      print(f'After time {self.trial_number}, {new_trials}: '
            f'Answer was {new_answer}, level is now {self.level}')
    return self.level

  def _response_rule(self):
    """Decides whether we change the level based on history."""
    last_run = self.run_trial_history[-1]
    # pylint: disable=g-explicit-length-test  Think of this as a list.
    if len(last_run) > 0:
      # Change level based on result of last test, down after positive result.
      self._change_level_down(last_run[-1])

  def _change_level_down(self, down):
    """Changes the level of an experiment and updates run history."""
    if self.debug:
      print('  Changing level down:', down)
    if self.multiplicative_step:
      if down:
        self.level /= (1 + self.change_delta)
      else:
        self.level *= (1 + self.change_delta)
    else:
      if down:
        self.level -= self.change_delta
      else:
        self.level += self.change_delta
    self.change_history.append(down)

    # Check to see if we are starting a new run
    if len(self.change_history
          ) > 1 and self.change_history[-2] != self.change_history[-1]:
      # Start a new run since direction changed
      if self.debug:
        print('Starting a new run...')
      self.run_trial_history.append([])
      if self.decrease_step_by_run:
        self.change_delta = self.initial_change_delta / len(
            self.run_trial_history)

  def calculate_threshold(self):
    start, end = self.run_boundaries(every_nth=1)
    num_runs = 2 * (len(start) // 2)  # Even number of runs
    start_levels = [self.level_history[i] for i in start[-num_runs:]]
    end_levels = [self.level_history[i] for i in end[-num_runs:]]
    all_points = start_levels + end_levels
    return sum(all_points) / float(num_runs) / 2.0

  def run_boundaries(self, every_nth=2):
    """Finds the boundaries between runs, defined as a stretch of ups or downs.

    Args:
      every_nth: Return every n'th boundary (so we can get every other one
        for plotting.)

    Returns:
      start, end: the starting and ending index for each run.  Note, the
      end of one trial is the start of the next.
    """

    # Add alternating color background for each run
    run_lengths = [len(sublist) for sublist in self.run_trial_history]
    run_boundaries = np.cumsum([0] + run_lengths)
    start = [i - 1 for i in run_boundaries[1::every_nth]]
    end = [i - 1 for i in run_boundaries[2::every_nth]]
    return start, end

  def plot_response(self, title: str = '') -> None:
    """Create a graph to summarize an experiment.

    Plot each test and its result, show vertical bars for each run.

    Args:
      title: Text to put at the top fo the graph.
    """
    # pylint: disable=g-complex-comprehension   Unpack list of lists
    run_history = [
        item for sublist in self.run_trial_history for item in sublist
    ]

    # Plot positive and negative trials separately
    positives = [i for i, v in enumerate(run_history) if v]
    negatives = [i for i, v in enumerate(run_history) if not v]
    # https://stackoverflow.com/questions/18272160/access-multiple-elements-of-list-knowing-their-index
    plt.plot(positives,
             operator.itemgetter(*positives)(self.level_history), '+')
    plt.plot(negatives,
             operator.itemgetter(*negatives)(self.level_history), '*')

    start, end = self.run_boundaries()
    for s, e in zip(start, end):
      plt.gca().axvspan(s, e, zorder=0, alpha=0.25)
    plt.title(title)
    plt.xlabel('Trial Number')
    plt.ylabel('Experiment Level')

  def results(self) -> Dict[str, Any]:
    return {
        'run_boundaries': self.run_boundaries,
        'run_trial_history': self.run_trial_history,
        'level_history': self.level_history,
        'experimental_threshold': self.calculate_threshold(),
    }


class Levitt2down1upAdditive(LevittExp):
  """A modification of the basic Levitt experiment.

  Now decreases the level after *two* positive results, and goes up after
  just one.
  """

  def _response_rule(self):
    last_run = self.run_trial_history[-1]
    new_trials = len(last_run)
    # pylint: disable=g-explicit-length-test  Think of this as a list.
    if len(last_run) > 0 and not last_run[-1]:  # Made a mistake
      self._change_level_down(False)
    elif len(last_run) > 1 and new_trials > 1 and last_run[-1] and last_run[-2]:
      # Got two right answers
      self._change_level_down(True)

File: extras/web/training/levitt_experiment.js
/**
 * Copyright 2021-2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Implements a basic version of the Transformed Up-Down Methods in
 * Psychoacoustics (Levitt 1970), adapted from psycho/levitt_experiment.py, for
 * use with web-based experiment tools. Handles recording results and tracking
 * runs, as well as calculating final threshold values.
 */

class LevittExperiment {
  /**
   * Constructs an experiment in initial state.
   */
  constructor(
      initialLevel, changeDelta, decreaseStepByRun = false,
      multiplicativeStep = false, maxRuns = 16, debug = false, minLevel = 1,
      maxLevel = 63, responseRuleNumber = 1) {
    if (changeDelta <= 0 || changeDelta > 10) {
      console.error('Change delta must greater than 0, and <= 10.');
    }

    if ((maxRuns % 2) || maxRuns < 1) {
      console.error('Max runs must be an even positive integer.');
    }

    /** @type {number} */ this.initialLevel = initialLevel;
    /** @type {number} */ this.level = initialLevel;
    /** @type {number} */ this.initialChangeDelta = changeDelta;
    /** @type {number} */ this.changeDelta = changeDelta;
    // Nested history of trials (run x trial #)
    /** @type {!Array} */ this.runTrialHistory = [[]];
    // Track change events to delineate runs
    /** @type {!Array} */ this.changeHistory = [];
    // History of all levels over time
    /** @type {!Array} */ this.levelHistory = [];
    /** @type {number} */ this.trialNumber = 0;
    /** @type {boolean} */ this.decreaseStepByRun = decreaseStepByRun;
    /** @type {boolean} */ this.multiplicativeStep = multiplicativeStep;
    /** @type {number} */ this.maxRuns = maxRuns;
    /** @type {number} */ this.minLevel = minLevel;
    /** @type {number} */ this.maxLevel = maxLevel;
    /** @type {number} */ this.responseRuleNumber = responseRuleNumber;
    /** @type {boolean} */ this.debug = debug;
  }

  /**
   * Provides current number of runs (length of runTrialHistory).
   * @return {number} Current number of runs.
   */
  get runNumber() {
    return this.runTrialHistory.length;
  }

  /**
   * Indicates whether experiment is ongoing based on current run.
   * @return {boolean} Experiment status.
   */
  get incomplete() {
    return this.runNumber <= this.maxRuns;
  }

  /**
   * Indicates whether the current level has saturated to the min or max.
   * @return {boolean} True if saturated.
   */
  get saturated() {
    return !(this.minLevel < this.level && this.level < this.maxLevel);
  }

  /**
   * Records the response to a trial, adjusts history, and sets new level.
   * @param {boolean} newAnswer Correctness of the user response.
   * @return {number} New level value.
   */
  noteResponse(newAnswer) {
    this.levelHistory.push(this.level);
    this.runTrialHistory[this.runNumber - 1].push(newAnswer);
    this.responseRule();
    this.trialNumber += 1;
    if (this.debug) {
      let lastRun = this.runTrialHistory[this.runNumber - 1];
      let newTrials = lastRun.length;
      console.log(
          'After time ' + this.trialNumber + ', ' + newTrials + ': ' +
          ' Answer was ' + newAnswer + ', level is now ' + this.level);
    }
    return this.level;
  }

  /**
   * Defines the response rule for whether to change the level.
   */
  responseRule() {
    switch (this.responseRuleNumber) {
      case 1:
        this.rule1();
        break;
      case 2:
        this.rule2();
        break;
      default:
        console.log('No Rule Defined');
    }
  }

  /**
   * Rule Levitt_1:
   *    If the run has entries, adjust level up after one incorrect
   *    response (-), adjust level down after one correct response (+).
   */
  rule1() {
    let lastRun = this.runTrialHistory[this.runNumber - 1];
    if (lastRun.length > 0) {
      let responseResult = lastRun[lastRun.length - 1];
      this.changeLevelDown(responseResult);
    }
  }

  /**
   * Rule Levitt_2:
   *    If the run has entries, adjust level up after one incorrect
   *    response (+ - or -), adjust level down after two sequential correct
   *    responses (+ +).
   */
  rule2() {
    let lastRun = this.runTrialHistory[this.runNumber - 1];
    // incorrect reponse:
    if (lastRun.length > 0 && !lastRun[lastRun.length - 1]) {
      this.changeLevelDown(false);
    } else if (lastRun.length > 1 && (lastRun.length % 2 == 0) &&
          lastRun[lastRun.length - 1] && lastRun[lastRun.length - 2]) {
      this.changeLevelDown(true);
    }
  }

  /**
   * Determines whether a new run should start, by comparing the last answer
   * to the current answer. If they differ, or if the level has saturated, begin
   * a new run.
   * @return {boolean} Whether to start a new run.
   */
  beginNewRun() {
    let changeLength = this.changeHistory.length;
    let lastChange = this.changeHistory[changeLength - 2];
    let thisChange = this.changeHistory[changeLength - 1];
    return (changeLength > 1) && (lastChange != thisChange || this.saturated);
  }

  /**
   * Limits the acceptable gains to within minimum and maximum bounds;
   * displays a warning message if bounds are violated.
   */
  boundsCheckGain() {
    if (this.level > this.maxLevel) {
      console.error('MAXIMUM GAIN ERROR');
      this.level = this.maxLevel;
    } else if (this.level < this.minLevel) {
      console.error('MINIMUM GAIN ERROR');
      this.level = this.minLevel;
    }
  }

  /**
   * Changes the level of an experiment and updates run history.
   * @param {boolean} down Whether the level should change down.
   */
  changeLevelDown(down) {
    if (this.debug) {
      console.log('  Changing level ', down ? 'down' : 'up');
    }

    if (this.multiplicativeStep) {
      if (down) {
        this.level /= (1 + this.changeDelta);
      } else {
        this.level *= (1 + this.changeDelta);
      }
    } else {
      if (down) {
        this.level -= this.changeDelta;
      } else {
        this.level += this.changeDelta;
      }
    }
    this.boundsCheckGain();
    this.changeHistory.push(down);

    if (this.beginNewRun()) {
      if (this.debug) {
        console.log('*******Starting a new run...********');
      }
      this.runTrialHistory.push([]);
      if (this.decreaseStepByRun) {
        this.changeDelta =
            this.initialChangeDelta / this.runTrialHistory.length;
      }
    }
  }

  /**
   * Finds the boundaries between runs, defined as a stretch of ups or downs.
   * @param {number} everyNth Step size between runs of interest.  Use 2
   *    for calculating the threshold using midpoint estimation.
   * @return {!Array<!Array<number>>} First entry is an array of start indices;
   *    second entry is an array of end indices.
   */
  runBoundaries(everyNth) {
    let runLengths = this.runTrialHistory.map((run) => run.length);
    let runBoundaries = [0];
    for (let i = 0; i < runLengths.length; i++) {
      runBoundaries.push(runBoundaries[i] + runLengths[i]);
    }
    // Excludes the first run since it has an arbitrary initial point.
    let startIndices = runBoundaries.slice(1, -1)
                           .filter((e, i) => (i % everyNth) === 0)
                           .map((e) => e - 1);
    let endIndices = runBoundaries.slice(2)
                         .filter((e, i) => (i % everyNth) === 0)
                         .map((e) => e - 1);
    return [startIndices, endIndices];
  }

  /**
   * Calculates a final threshold value as the mean of the midpoints of
   * every second run, over an even number of runs.
   * (see Fig. 4, Levitt 1970)
   * @return {number} Threshold value.
   */
  calculateThreshold() {
    let runBoundaries = this.runBoundaries(2);
    let starts = runBoundaries[0];
    let ends = runBoundaries[1];
    let startLevels = starts.map((i) => this.levelHistory[i]);
    let endLevels = ends.map((i) => this.levelHistory[i]);
    let sumStarts = (startLevels.reduce((a, b) => a + b));
    let sumEnds = (endLevels.reduce((a, b) => a + b));
    let mean = (sumStarts + sumEnds) / starts.length / 2;
    return mean;
  }

  /**
   * Creates a readable string of all the experiment data.
   * @return {string} Experiment data for download or display.
   */
  getHistory() {
    let contents = '';
    contents += 'Trials: ' + this.runTrialHistory.toString() + '\n';
    contents += 'Levels: ' + this.levelHistory.toString() + '\n';
    return contents;
  }
}
